template_test.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"testing"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBotMessage(t *testing.T) {
+	tt := []struct {
+		name            string
+		input           BotMessageData
+		expectedMessage string
+		expectedError   error
+	}{
+		{
+			name: "valid template",
+			input: BotMessageData{
+				Template: "'{{.Branch}}' will auto-release to: {{range .AutoReleaseEnvironments}}\n {{.}}{{end}}",
+				Branch:   "master",
+				AutoReleaseEnvironments: []string{
+					"dev",
+				},
+			},
+			expectedMessage: "'master' will auto-release to: \n dev",
+			expectedError:   nil,
+		},
+		{
+			name: "invalid template",
+			input: BotMessageData{
+				Template: "'{{.Branch}}' will auto-release to: {{range .AutoReleaseEnvironments}}\n {{.}}",
+				Branch:   "master",
+				AutoReleaseEnvironments: []string{
+					"dev",
+				},
+			},
+			expectedMessage: "",
+			expectedError:   errors.New(""),
+		},
+		{
+			name: "no template",
+			input: BotMessageData{
+				Template: "",
+				Branch:   "master",
+				AutoReleaseEnvironments: []string{
+					"dev",
+				},
+			},
+			expectedMessage: "",
+			expectedError:   errors.New(""),
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			// Arrange
+			var actualMessage string
+			var actualError error
+
+			// Act
+			actualMessage, actualError = BotMessage(tc.input)
+
+			// Assert
+			assert.Equal(t, tc.expectedMessage, actualMessage)
+
+			if tc.expectedError != nil {
+				assert.Error(t, actualError)
```

Comment:
This is only an assertion as to the existence of an error, not that it matches what the test case expects. The following change will added the equality of errors.
Notice that your test cases properly fails with this change as they all expect an empty error. 🙈 

```suggestion
				assert.EqualError(t, tc.expectedError, actualError)
```

Reply:
But but... I don't know how to recreate the errors which is returned in the uut 🙉

Reply:
Either we expose the error as a sentinal error value like this.

```go
var errTemplateEmpty = errors.New("template empty")

func BetMessage(m BotMessageData{}) (string, error) {
  if m.Template == "" {
    return "", errTemplateEmpty
  }
  // ...
}
```

This way they can be referenced in tests the same way.

Otherwise you can create an equal error like `errors.New("template empty")` in the test case.

Reply:
Thanks